src/core/bonded_interactions/angle_cosine.hpp

```diff
@@ -100,9 +92,10 @@ inline bool calc_angle_cosine_force(Particle const *p_mid,
  *  @param[out] _energy   Energy.
  *  @retval false
  */
-inline bool angle_cosine_energy(Particle const *p_mid, Particle const *p_left,
-                                Particle const *p_right,
-                                Bonded_ia_parameters const *iaparams,
+inline bool angle_cosine_energy(Particle const *const p_mid,
```

Comment:
why do we have this strange function signature for all those bonded potentials? if a bond can break, can we just handle this within the calculation and instead return the energy. If we want to keep the structure, we could also return a ```boost::optional```

Reply:
We need to forward the information about broken bonds to the calling function, which will produce a runtime error message that will be gathered on the master MPI node to stop the simulation (I think). In C, I would personally prefer having the broken bond as a side-effect of the function, i.e. `inline double angle_cosine_energy(..., bool *bond_broken)`. We're going to rewrite these signatures as `std::tuple<double, bool>` soon, and probably also remove the bool in those potentials where bonds can't break.

Reply:
Seems like the body of this conditional cannot be reached because the same check is performed earlier to exit the function prematurely:
https://github.com/espressomd/espresso/blob/de20f8e94cc5a69adf3ab916e92b23cf49cc3032/src/core/energy_inline.hpp#L348
The force and energy functions probably need some restructuring...

Reply:
Actually, since the energy is undefined for a broken bond, we could consider returning `std::nan` as a value and have a NaN check trigger the runtime error message. This could also be done in force kernels. What do you think?

Reply:
Why do we need to forward this information at all? If we encounter a broken bond, we can write to ```runtimeErrorMessage``` directly in the function.

Concerning the NaN approach:
CPP reference:

A common use case for optional is the return value of a function that may fail. As opposed to other approaches, such as std::pair<T,bool>, optional handles expensive-to-construct objects well and is more readable, as the intent is expressed explicitly.

Thats exactly our usecase, right?  Why not use the boost alternative to that?

Reply:
that would also be more expressive, since the caller can see from the return value that the calculation my fail

Reply:
+1 for optional.

Reply:
Just tried `std::nan`, `std::optional` (C++17) and `boost::optional` on a toy example and all give the exact same assembly code. The boost version however adds a lot of extra symbols (objdump: 3420 lines, vs. 832 for the `std::nan`/`std::optional` solutions), is that fine?

Reply:
When reproducing the structure of `energy_inline.hpp`, the `std::optional`/`boost::optional` are identical and more efficient than `std::nan`: [godbolt-NaN](https://godbolt.org/z/su_sK-), [godbolt-optional](https://godbolt.org/z/ues41l), [godbolt-boost](https://godbolt.org/z/o2HzDl)

Reply:
Let's just make it an optional then.

Reply:
Regarding your example: I think if you have already check if the optional has a value you should use `boost::optional::get` for value extraction, instead of value. This makes all
of the exception [code go away](https://godbolt.org/z/iWQ07w) because it can never throw.

Reply:
Looks much better! I think this refactoring should be delayed to the PR where we move from output parameters to tuples, e.g.
```c++
boost::optional<double> tab_angle_energy(...);
boost::optional<std::tuple<Vector3d, Vector3d, Vector3d>> calc_tab_angle_force(...);
```
I already have the tuple version of the code in an old commit, I just need to remove the bool and add the boost::optional around each tuple. And fix a bug in that old commit.

Reply:
The tuple version without `boost::optional` ([fix-2502-p1b...fix-2502-p2](https://github.com/jngrad/espresso/compare/fix-2502-p1b...fix-2502-p2)) is now passing all tests, but the LJ script from the benchmarks shows it's 23% slower. The creation of all these intermediate `Vector3d` is probably the culprit. I'll investigate that and check if `boost::optional` has any effect on performance.